src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c

```diff
@@ -771,14 +821,16 @@ static int32_t ComplexEndsWith(const UCollator* pCollator, UErrorCode* pErrorCod
         int32_t idx = usearch_last(pSearch, pErrorCode);
         if (idx != USEARCH_DONE)
         {
-            if ((idx + usearch_getMatchedLength(pSearch)) == patternLength)
+            int32_t matchEnd = idx + usearch_getMatchedLength(pSearch);
+            assert(matchEnd <= textLength);
+
+            if (matchEnd == textLength)
             {
-                result = TRUE;
+                return TRUE;
```

Comment:
I think that if you return from here the `usearch_close` method is not going to be executed:

https://github.com/dotnet/runtime/blob/880fb633956aafbd7ddcd49f48328f0d3b69840f/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c#L839

Reply:
Whoops. Good catch!

Reply:
This never would've happened had we stuck with RAII principles. ;)

Reply:
> RAII 

Why would anyone need `try`, `catch` and `finally` when you can just use nested `if`s and error codes? :D

Perhaps we should switch from `C` to `C++`?

Reply:
We explicitly switched from C++ to C, which mono needs.
https://github.com/dotnet/coreclr/pull/22378

Reply:
@stephentoub thanks for providing the link to explanation!